### PR TITLE
Added Kafka as Transport

### PR DESCRIPTION
**Pull request by vMaroon**

### CODE REVIEW

File: cmd/manager/main.go [hunk 1]
```diff
@@ -39,35 +49,71 @@ func printVersion(log logr.Logger) {
...
-	log := ctrl.Log.WithName("cmd")
-
-	printVersion(log)
+// function to choose transport type based on env var.
+func getTransport(transportType string) (transport.Transport, error) {
+	switch transportType {
+	case kafkaTransportTypeName:
+		kafkaProducer, err := kafkaClient.NewHOHProducer(ctrl.Log.WithName("kafka-client"))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create hoh-kafka-producer: %w", err)
```
Comment (nirrozenbaum) [inline cmd/manager/main.go:58]: no need for specifying "hoh"

File: cmd/manager/main.go [hunk 2]
```diff
@@ -39,35 +49,71 @@ func printVersion(log logr.Logger) {
...
+		return nil, errEnvVarIllegalValue
+	}
+}
 
+func readEnvVars(log logr.Logger) (string, time.Duration, string, error) {
 	leaderElectionNamespace, found := os.LookupEnv(envVarControllerNamespace)
 	if !found {
 		log.Error(nil, "Not found:", "environment variable", envVarControllerNamespace)
-		return 1
+		return "", 0, "", errEnvVarNotFound
```
Comment (nirrozenbaum) [inline cmd/manager/main.go:78]: in all returning fields, in the places where we have error, readEnvVars shouldn't log the error.
it should just return wrapped error using fmt.Errorf(... %w, err).
main function should log the error and exit.

so no need to give this function the logger as an argument.
see for example here:
https://github.com/open-cluster-management/hub-of-hubs-spec-transport-bridge/blob/main/pkg/transport/sync-service/sync_service.go#L53

File: cmd/manager/main.go [hunk 3]
```diff
@@ -39,35 +49,71 @@ func printVersion(log logr.Logger) {
...
+	pflag.Parse()
+
+	ctrl.SetLogger(zap.Logger())
+	log := ctrl.Log.WithName("cmd")
+
+	printVersion(log)
+
+	leaderElectionNamespace, syncInterval, transportType, err := readEnvVars(log)
+	if err != nil {
+		return 1
```
Comment (nirrozenbaum) [inline cmd/manager/main.go:115]: log error, then exit

File: deploy/hub-of-hubs-spec-transport-bridge.yaml.template [hunk 4]
```diff
@@ -71,6 +71,14 @@ spec:
...
                   name: hub-of-hubs-database-transport-bridge-secret
                   key: url
+            - name: HOH_TRANSPORT_TYPE
+              value: "$HOH_TRANSPORT_TYPE"
+            - name: KAFKA_PRODUCER_ID
+              value: "hoh-spec-producer"
+            - name: KAFKA_PRODUCER_HOSTS
+              value: "$KAFKA_PRODUCER_HOSTS"
+            - name: KAFKA_PRODUCER_TOPIC
+              value: "spec-path-topic"
```
Comment (nirrozenbaum) [inline deploy/hub-of-hubs-spec-transport-bridge.yaml.template:81]: let's try not to repeat things.
we can set the topic name to `Spec`.
no need for setting topic with suffix topic :)

File: deploy/hub-of-hubs-spec-transport-bridge.yaml.template [hunk 5]
```diff
@@ -71,6 +71,14 @@ spec:
                 secretKeyRef:
                   name: hub-of-hubs-database-transport-bridge-secret
                   key: url
+            - name: HOH_TRANSPORT_TYPE
+              value: "$HOH_TRANSPORT_TYPE"
+            - name: KAFKA_PRODUCER_ID
+              value: "hoh-spec-producer"
+            - name: KAFKA_PRODUCER_HOSTS
+              value: "$KAFKA_PRODUCER_HOSTS"
```
Comment (nirrozenbaum) [inline deploy/hub-of-hubs-spec-transport-bridge.yaml.template:79]: see previous comments, this can be set to the cluster internal service, no need for setting env var.

File: deploy/hub-of-hubs-spec-transport-bridge.yaml.template [hunk 6]
```diff
@@ -71,6 +71,14 @@ spec:
                 secretKeyRef:
                   name: hub-of-hubs-database-transport-bridge-secret
                   key: url
+            - name: HOH_TRANSPORT_TYPE
+              value: "$HOH_TRANSPORT_TYPE"
+            - name: KAFKA_PRODUCER_ID
+              value: "hoh-spec-producer"
```
Comment (nirrozenbaum) [inline deploy/hub-of-hubs-spec-transport-bridge.yaml.template:77]: not need for repeating the word producer.
can be set to `hub-of-hubs`

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
+package kafkaclient
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/go-logr/logr"
+	kclient "github.com/open-cluster-management/hub-of-hubs-kafka-transport/kafka-client/kafka-producer"
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:10]: why not call the create alias to the package as "producer"?
it's not so clear what is kclient.

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/open-cluster-management/hub-of-hubs-spec-transport-bridge/pkg/transport"
+)
+
+// NewHOHProducer returns a new instance of HOHProducer object.
+func NewHOHProducer(log logr.Logger) (*HOHProducer, error) {
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:15]: this function violates best practices.
you should first try to create the kafka producer. only if it succeeds and there are no errors you should create the HoHProducer. so your function should look something like the following:

```
deliveryChan := make(chan kafka.Event)

kafkaProducer, err := kclient.NewKafkaProducer(deliveryChan)
if err != nil {
	return nil, fmt.Errorf("failed to create kafka producer: %w", err)
}

return &HoHProducer{
		deliveryChan:  deliveryChan,
		stopChan:      make(chan struct{}, 1),
		kafkaProducer: kafkaProducer,
		log:           log,
}, nil
```

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+		return nil, fmt.Errorf("failed to create kafka producer: %w", err)
+	}
+
+	kp.kafkaProducer = kafkaProducer
+
+	return kp, nil
+}
+
+// HOHProducer abstracts hub-of-hubs-kafka-transport kafka-producer's generic usage.
+type HOHProducer struct {
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:34]: let's rename to HoHProducer.

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+	kafkaProducer *kclient.KafkaProducer
+	deliveryChan  chan kafka.Event
+	stopChan      chan struct{}
+	startOnce     sync.Once
+	stopOnce      sync.Once
+}
+
+// deliveryHandler handles results of sent messages.
+// For now failed messages are only logged.
+func (p *HOHProducer) deliveryHandler(e *kafka.Event) {
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:45]: rename `p` to `producer`, helps for readability

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	p.stopOnce.Do(func() {
+		p.kafkaProducer.Close()
+		p.stopChan <- struct{}{}
+		close(p.stopChan)
+		close(p.deliveryChan)
+	})
+}
+
+// SendAsync sends a message to the sync service asynchronously.
+func (p *HOHProducer) SendAsync(id string, msgType string, version string, payload []byte) {
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:94]: if we want to keep the same structure as sync service the sendAsync function can only push to a msgChan channel and a different go routine "distributeMessages" should read from the channel and distribute it. 

what will happen if kafkaProducer is not responding? will your code hang?
if that's the case than it's a must to separate to different functions. 
we can't say we sendAsync but then hang.

if it will not hang, it's not a necessary step but still would be nicer to have same structured code.

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+	message := &transport.Message{
+		ID:      id,
+		MsgType: msgType,
+		Version: version,
+		Payload: payload,
+	}
+
+	bs, err := json.Marshal(message)
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:102]: var name not readable, let's call the child with his name:
`messageBytes` instead of `bs`.

File: cmd/manager/main.go [hunk 13]
```diff
@@ -39,35 +49,71 @@ func printVersion(log logr.Logger) {
...
-
-	ctrl.SetLogger(zap.Logger())
-	log := ctrl.Log.WithName("cmd")
-
-	printVersion(log)
+// function to choose transport type based on env var.
+func getTransport(transportType string) (transport.Transport, error) {
+	switch transportType {
+	case kafkaTransportTypeName:
+		kafkaProducer, err := kafkaClient.NewHOHProducer(ctrl.Log.WithName("kafka-client"))
```
Comment (nirrozenbaum) [inline cmd/manager/main.go:56]: no need for the word "client". just kafka.

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+	if err != nil {
+		p.log.Error(err, "Failed to send message", "Message ID", message.ID)
+		return
+	}
+
+	err = p.kafkaProducer.ProduceAsync(&bs)
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:108]: I adopted the one line command + err check where possible.
I think this is the recommended way in Go (again, only when it's possible)
here it's possible using the following:
```
if err = p.kafkaProducer.ProduceAsync(&messageBytes); err != nil {
    ....
}
```

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+	if err != nil {
+		p.log.Error(err, "Failed to send message", "Message ID", message.ID)
+	}
+}
+
+// GetVersion returns an empty string if the object doesn't exist or an error occurred.
+func (p *HOHProducer) GetVersion(id string, msgType string) string {
+	// TODO: implement with consumer
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:116]: I'm wondering.. is there a way to implement this elegantly?
if there is maybe you can implement it and push this one as well..
not a big deal if it will be done later, just making sure we won't forget about it.

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+	deliveryChan  chan kafka.Event
+	stopChan      chan struct{}
+	startOnce     sync.Once
+	stopOnce      sync.Once
+}
+
+// deliveryHandler handles results of sent messages.
+// For now failed messages are only logged.
+func (p *HOHProducer) deliveryHandler(e *kafka.Event) {
+	switch ev := (*e).(type) {
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:46]: variable names here are not readable and therefore this function is very hard to follow even though it's only 10 lines.

we need to better name the variables. instead of `e` use kafkaEvent.
instead of `ev` use kafkaEventType or something like that.

**Review by nirrozenbaum — COMMENTED** [review]
@vMaroon very nice code, clean and well written.

though I wrote here a lot of comments, all of them are minor and can be fixed very quickly.
excellent work, well done!

let's get these points fixed quickly and merge this PR.

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/go-logr/logr"
+	kclient "github.com/open-cluster-management/hub-of-hubs-kafka-transport/kafka-client/kafka-producer"
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:11]: alias to kafka instead of kclient, as in other places..

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+
+// deliveryHandler handles results of sent messages.
+// For now failed messages are only logged.
+func (p *Producer) deliveryHandler(kafkaEvent *kafka.Event) {
+	switch event := (*kafkaEvent).(type) {
+	case *kafka.Message:
+		if event.TopicPartition.Error != nil {
+			load := &transport.Message{}
+
+			err := json.Unmarshal(event.Value, load)
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:53]: one line check:
```
if err := json.Unmarshal(event.Value, load); err != nil { 
...
```

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+			if err != nil {
+				p.log.Error(err, "Failed to deliver message",
+					"topic name", event.TopicPartition)
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:56]: keep in same line unless linter is sad

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+	deliveryChan  chan kafka.Event
+}
+
+// deliveryHandler handles results of sent messages.
+// For now failed messages are only logged.
+func (p *Producer) deliveryHandler(kafkaEvent *kafka.Event) {
+	switch event := (*kafkaEvent).(type) {
+	case *kafka.Message:
+		if event.TopicPartition.Error != nil {
+			load := &transport.Message{}
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:51]: why `load` and not `message`?
you're unmarshaling transport.Message..

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+	if err = p.kafkaProducer.ProduceAsync(messageBytes, []byte(id), []byte(msgType), []byte(version)); err != nil {
+		p.log.Error(err, "Failed to send message", "Message ID", message.ID)
+	}
+}
+
+// GetVersion returns an empty string if the object doesn't exist or an error occurred.
+func (p *Producer) GetVersion(id string, msgType string) string {
+	// we assume that the spec transport bridge is always running, and at worst was offline for generousDownTime.
+	// we also assume that the producer's writes succeeded in previous runs.
+	return time.Now().Add(-1 * generousDownTime).Format(timeFormat)
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:112]: leave empty with TODO...

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+func (p *Producer) handleDelivery(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case e := <-p.deliveryChan:
```
Comment (nirrozenbaum) [inline pkg/transport/kafka-client/kafka_producer_client.go:121]: event or kafkaEvent instead of e

File: pkg/transport/kafka-client/kafka_producer_client.go [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+		Payload: payload,
+	}
+
+	messageBytes, err := json.Marshal(message)
+	if err != nil {
+		p.log.Error(err, "Failed to send message", "Message ID", message.ID)
+		return
+	}
+
+	if err = p.kafkaProducer.ProduceAsync(messageBytes, []byte(id), []byte(msgType), []byte(version)); err != nil {
```
[the remark this answers is not in the document]
Comment (vMaroon) [reply pkg/transport/kafka-client/kafka_producer_client.go]: I would rather keep kafka's repo's produce API like this as the real Produce() call also works this way.
All other comments were executed.

File: pkg/transport/kafka-client/kafka_producer_client.go (hunk 17 above)
Comment (vMaroon) [reply pkg/transport/kafka-client/kafka_producer_client.go]: It is not possible to do that without changing confluent's import into something else (it is currently kafka) and it feels wrong to give it any other name, since it still provides exports such as kafka.Message, kafka.Event...
"kclient" was changed to "kafkaClient" as basically that is what the kafka repo wraps and provides.
